src/adapter/src/coord.rs

```diff
@@ -649,23 +661,113 @@ impl<S: Append + 'static> Coordinator<S> {
+            let mut compute_policy_updates = Vec::new();
+            for id in compute_ids.iter() {
+                let policy = match compaction_window_ms {
+                    Some(time) => ReadPolicy::lag_writes_by(time),
+                    None => ReadPolicy::ValidFrom(Antichain::from_elem(Timestamp::minimum())),
+                };
+
+                let mut read_capability: ReadCapability<_> = policy.into();
+
+                if let Some(timeline) = self.get_timeline(*id) {
+                    let TimelineState { read_holds, .. } =
+                        self.ensure_timeline_state(timeline).await;
+                    read_capability
+                        .holds
+                        .update_iter(Some((read_holds.time, 1)));
+                    read_holds
+                        .id_bundle
+                        .compute_ids
+                        .entry(*compute_instance)
+                        .or_default()
+                        .insert(*id);
+                }
+
+                self.read_capability.insert(*id, read_capability);
+                compute_policy_updates.push((*id, self.read_capability[&id].policy()));
+            }
+            self.controller
+                .compute_mut(*compute_instance)
+                .unwrap()
+                .set_read_policy(compute_policy_updates)
+                .await
+                .unwrap();
+        }
+
+        let mut storage_policy_updates = Vec::new();
+        for id in id_bundle.storage_ids.iter() {
             let policy = match compaction_window_ms {
                 Some(time) => ReadPolicy::lag_writes_by(time),
                 None => ReadPolicy::ValidFrom(Antichain::from_elem(Timestamp::minimum())),
             };
-            self.read_capability.insert(id, policy.clone().into());
-            policy_updates.push((id, self.read_capability[&id].policy()));
+
+            let mut read_capability: ReadCapability<_> = policy.into();
+
+            if let Some(timeline) = self.get_timeline(*id) {
+                let TimelineState { read_holds, .. } = self.ensure_timeline_state(timeline).await;
+                read_capability
+                    .holds
+                    .update_iter(Some((read_holds.time, 1)));
+                read_holds.id_bundle.storage_ids.insert(*id);
+            }
+
+            self.read_capability.insert(*id, read_capability);
+            storage_policy_updates.push((*id, self.read_capability[&id].policy()));
         }
         self.controller
-            .compute_mut(instance)
-            .unwrap()
-            .set_read_policy(policy_updates)
+            .storage_mut()
+            .set_read_policy(storage_policy_updates)
             .await
             .unwrap();
     }
 
+    /// Ensures that a global timeline state exists for `timeline`.
+    async fn ensure_timeline_state(&mut self, timeline: Timeline) -> &mut TimelineState<Timestamp> {
+        if !self.global_timelines.contains_key(&timeline) {
```

Comment:
`self.global_timelines.entry(timeline).or_insert_with(|| ...)` should work here

Reply:
Unfortunately not because the `await` isn't allowed in the closure. Unless there's a way to call async functions in the closure.